cli/lib/kontena/cli/stacks/yaml/reader.rb

```diff
@@ -368,12 +368,14 @@ def interpolate(content, use_opto: true, substitutions: {}, raise_on_unknown: fa
               if use_opto
                 opt = variables.option(var)
                 if opt.nil?
-                  raise RuntimeError, "Undeclared variable '#{var}' in #{file}:#{line_num} -- #{row}" if raise_on_unknown
+                  to_env = variables.find { |opt| Array(opt.to[:env]).include?(var) }
```

Comment:
Might want to consider a `Hash` for variable lookups at some point, this is `O(N*M)` on the number of variables/interpolations. With a complex enough stack you might notice it... needs profiling before it's worth considering trying to optimize it though.

BTW: Looks like the same also applies to `Opto::Group#option`...

Reply:
Well opto could just use [sets](http://ruby-doc.org/stdlib-2.4.1/libdoc/set/rdoc/Set.html) (which i believe to use hash internally). They're much faster than arrays and mostly behave in a similar fashion.